### PR TITLE
Particle plugin

### PR DESCRIPTION
**Pull request by davidgs**
### Required for all PRs:

- [X ] Signed [CLA](https://influxdata.com/community/cla/).
- [ X] Associated README.md updated.
- [X ] Has appropriate unit tests.

This Plugin adds the particle.io webhooks service as an Input to Telegraf. It takes a JSON object as input of the form:

```
{
	  "event": "measurement_name",
	  "data": {
		  "tags": {
			  "tag_1": "tag_1_value",
			  "tag_2": "tag_2_value"
		  },
		  "values": {
			  "name": value,
			  "name": value
		  }
	  },
	  "published_at": "2017-09-28T21:54:10.897Z",
	  "coreid": "123456789938323536",
	  "userid": "1234ee123ac8e5ec1231a123d",
	  "influx_db": "database_name"
  }
```
The cored, userid, etc. are automatically supplied by the Particle.io cloud. The "influx_db" parameter must be supplied as additional JSON when defining the Webhook in the Particle.io Cloud.

### CODE REVIEW

File: plugins/inputs/webhooks/particle/README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# particle webhooks
+
+You should configure your Particle.io's Webhooks to point at the `webhooks` service. To do this go to `console.particle.io/` and click `Integrations > New Integration > Webhook`. In the resulting page set `URL` to `http://<my_ip>:1619/particle`, and  under `Advanced Settings` click on `JSON` and add:
```
Comment (danielnelson) [inline plugins/inputs/webhooks/particle/README.md]: Turn `console.particle.io/` into a url: `[name of link](https://console.particle.io/)`

File: plugins/inputs/webhooks/particle/particle_webhooks.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	return &event{
+		Data: data{
+			Tags:   make(map[string]string),
+			Fields: make(map[string]interface{}),
+		},
+	}
+}
+
+func (e *event) Time() (time.Time, error) {
+	return time.Parse("2006-01-02T15:04:05Z", e.PublishedAt)
```
Comment (danielnelson) [inline plugins/inputs/webhooks/particle/particle_webhooks.go]: Should be able to use time.RFC3339 as the layout.

File: plugins/inputs/webhooks/particle/particle_webhooks.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	if err := json.NewDecoder(r.Body).Decode(e); err != nil {
+		log.Println(err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	pTime, err := e.Time()
+	if err != nil {
+		pTime = time.Now()
+		log.Printf("error parsing particle event time: %s. Using telegraf host time instead: %s", e.PublishedAt, pTime)
```
Comment (danielnelson) [inline plugins/inputs/webhooks/particle/particle_webhooks.go]: What sort of situations would this occur in?  Maybe it should just be a hard 400 error?

File: plugins/inputs/webhooks/particle/particle_webhooks_test.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+	tags := map[string]string{
+		"id":       "230035001147343438323536",
+		"location": "TravelingWilbury",
+	}
+
+	acc.AssertContainsTaggedFields(t, "temperature", fields, tags)
+}
+
+func TestUnknowItem(t *testing.T) {
```
Comment (danielnelson) [inline plugins/inputs/webhooks/particle/particle_webhooks_test.go]: typo: TestUnknowItem -> TestUnknownItem

File: plugins/inputs/webhooks/particle/particle_webhooks_test.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	  "userid": "1234ee123ac8e5ec1231a123d",
+	  "version": 10,
+	  "public": false,
+	  "productID": 1234,
+	  "name": "sensor",
+	  "influx_db": "mydata"
+  }`
+}
+
+func UnknowJSON() string {
```
Comment (danielnelson) [inline plugins/inputs/webhooks/particle/particle_webhooks_test.go]: typo: UnknowJSON -> UnknownJSON

**Comment by danielnelson** [comment]
Tests on circle should pass if you rebase against master.

File: plugins/inputs/webhooks/particle/particle_webhooks_test.go (hunk 4 above)
Comment (davidgs) [reply plugins/inputs/webhooks/particle/particle_webhooks_test.go]: Typo carried over from the Rollbar plugin. I thought maybe it was historically required. :-)

Comment (danielnelson) [reply plugins/inputs/webhooks/particle/particle_webhooks_test.go]: lol

**Comment by davidgs** [comment]
So it fails CircleCI, but not because of the telegraf code, but because of a docker/aerospike problem.

**Comment by nhaugo** [comment]
@davidgs did you rebase?

**Comment by davidgs** [comment]
I did rebase. If you look at the CircleCI output, the tests on the Telegraf plugin webhooks/particle pass. It's a whole bunch of other stuff in Circle CI that fails.

**Comment by davidgs** [comment]
Let's start this all over from a clean slate ...

**Pull request closed without merging**